core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -148,78 +165,117 @@ private[kafka] object LogValidator {
-
-    var maxTimestamp = Record.NO_TIMESTAMP
-    val expectedInnerOffset = new LongRef(0)
-    val validatedRecords = new mutable.ArrayBuffer[Record]
-
-    records.deepEntries(true).asScala.foreach { logEntry =>
-      val record = logEntry.record
-      validateKey(record, compactedTopic)
-
-      if (record.magic > Record.MAGIC_VALUE_V0 && messageFormatVersion > Record.MAGIC_VALUE_V0) {
-        // Validate the timestamp
-        validateTimestamp(record, now, messageTimestampType, messageTimestampDiffMaxMs)
-        // Check if we need to overwrite offset, no in place assignment situation 3
-        if (logEntry.offset != expectedInnerOffset.getAndIncrement())
-          inPlaceAssignment = false
-        if (record.timestamp > maxTimestamp)
-          maxTimestamp = record.timestamp
+
+  def validateMessagesAndAssignOffsetsCompressed(records: MemoryRecords,
+                                                 offsetCounter: LongRef,
+                                                 currentTimestamp: Long,
+                                                 sourceCodec: CompressionCodec,
+                                                 targetCodec: CompressionCodec,
+                                                 compactedTopic: Boolean = false,
+                                                 messageFormatVersion: Byte = RecordBatch.CURRENT_MAGIC_VALUE,
+                                                 messageTimestampType: TimestampType,
+                                                 messageTimestampDiffMaxMs: Long): ValidationAndOffsetAssignResult = {
+      // Deal with compressed messages
+      // We cannot do in place assignment in one of the following situations:
+      // 1. Source and target compression codec are different
+      // 2. When magic value to use is 0 because offsets need to be overwritten
+      // 3. When magic value to use is above 0, but some fields of inner messages need to be overwritten.
+      // 4. Message format conversion is needed.
+
+      // No in place assignment situation 1 and 2
+      var inPlaceAssignment = sourceCodec == targetCodec && messageFormatVersion > RecordBatch.MAGIC_VALUE_V0
+
+      var maxTimestamp = RecordBatch.NO_TIMESTAMP
+      val expectedInnerOffset = new LongRef(0)
+      val validatedRecords = new mutable.ArrayBuffer[Record]
+
+      for (batch <- records.batches.asScala) {
+        ensureNonTransactional(batch)
+
+        for (record <- batch.asScala) {
+          if (!record.hasMagic(batch.magic))
+            throw new InvalidRecordException(s"Log record magic does not match outer magic ${batch.magic}")
+
+          record.ensureValid()
+          ensureNotControlRecord(record)
+          validateKey(record, compactedTopic)
+
+          if (!record.hasMagic(RecordBatch.MAGIC_VALUE_V0) && messageFormatVersion > RecordBatch.MAGIC_VALUE_V0) {
+            // No in place assignment situation 3
```

Comment:
Not sure why this comment was moved here. I think it made more sense alongside the check that sets `inPlaceAssignment = false`.